### ISSUE TITLE
Process.stdin.add() does not always send all data to the process on Windows before completing (even with flush())

### ISSUE THREAD

**Issue by mtolmacs** [dart-lang/sdk#50904]
This issue is related to the input size, so I provided an example `test.txt` file with a length where the issue triggers. If the stdin.add() is used with multiple smaller chunks sometimes I can get it to write the whole file out, sometimes only the multiples of the chunk size is written out and the last chunk is missing.

Code to reproduce:
```
import 'dart:io';
import 'dart:typed_data';

import 'package:path/path.dart';

final String workdir = dirname(Platform.script.toFilePath());

Future<int> main(List<String> arguments) async {
  final Uint8List bytes = File('$workdir/test.txt').readAsBytesSync();

  final Process proc = await Process.start(
    'dart',
    <String>[
      '$workdir/writetofile.dart',
      'debug.bin',
    ],
    runInShell: true,
    workingDirectory: workdir,
  );

  proc.stdin.add(bytes);
  stdout.addStream(proc.stdout);
  stderr.addStream(proc.stderr);

  await proc.stdin.flush();
  await proc.stdin.close();

  return 0;
}
```

Support script to write stdin to file the above script calls:
```
import 'dart:io';

Future<int> main(List<String> arguments) async {
  final File outputFile = File(arguments[0]);
  outputFile.createSync();
  final IOSink output = outputFile.openWrite();

  await output.addStream(stdin);
  await output.flush();
  await output.close();

  return 0;
}
```

The test input file:
[test.txt](https://github.com/dart-lang/sdk/files/10353004/test.txt)

Expected: The newly created debug.bin file has the exact same size the source `test.txt` file has.
Actual: The newly created debug.bin file has zero size.

Dart version: `Dart SDK version: 2.18.6 (stable) (Tue Dec 13 21:15:14 2022 +0000) on "windows_x64"`
Platform: Windows

Edit: The issue came from working with ProcessPools and WorkerJobs. The ProcessPool uses LocalProcessManager by default, which uses Process.start() and feeds the stdin asynchronously.

**Comment by cbracken** [comment]
@brianquinlan any thoughts on when you may be able to get to this? On the flutter side, https://github.com/flutter/engine/pull/36123 is blocked on a fix for this issue. (Desktop triage)

**Comment by brianquinlan** [comment]
I can reproduce this on Windows. Trying to figure it out now.

**Comment by brianquinlan** [comment]
Uh oh, adding a delay before the `await proc.stdin.close()` causes this to succeed i.e.

```dart
  ...
  await proc.stdin.flush();
  sleep(Duration(seconds: 2));
  await proc.stdin.close();
  ...
```

So I'm going to start by trying to figure out if `flush()` is doing it's job.

**Comment by brianquinlan** [comment]
Some things that I discovered:
1. the issue disappears if the pipe used by the pipe is larger than the size of the file i.e. 
```
        CreateNamedPipeW(pipe_name, PIPE_ACCESS_OUTBOUND | FILE_FLAG_OVERLAPPED,
                         PIPE_TYPE_BYTE | PIPE_WAIT,
                         1,     // Number of pipes
+                        4096,
+                        4096,
-                         1024,  // Out buffer size
-                         1024,  // In buffer size
                         0,     // Timeout in ms
                         NULL);
```
2. `flush` doesn't seem to do anything. The call to `add` makes an asynchronous call to `WriteFile`, which finishes after `flush` returns but before the call to `close`.
3. Given (2), there may be an issue where closing the pipe before the client has read the data results in the data being lost. The documentation is not clear on that issue. There is (`FlushFileBuffers`)[https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-flushfilebuffers), which blocks until the client reads the data.
4. There doesn't seem to be a problem with the `writetofile.dart` - if I replace it with a Python version than it also doesn't work. For completeness I should probably also test with a Python process invoker.

**Comment by mtolmacs** [comment]
This might be an issue with how and when the kernel flushes the data after `CloseHandle`. Since now there is no `FlushFileBuffers` before `CloseHandle`, the kernel sits on the last buffered 'page'  to be flushed later, but the receiving process already terminated when the `CloseHandle` on the pipes reaches the target process. Thus the delayed named pipe write is never received. This is not an issue if the target is a real file, not a pipe where the program listening on the other end terminates on std close.

**Comment by mtolmacs** [comment]
I just checked, I can confirm that adding `FlushFileBuffers(handle_);` to [Handle::DoClose()](https://github.com/dart-lang/sdk/blob/3571b97b24a2d8b27038284d5f0be58bd0733f61/runtime/bin/eventhandler_win.cc#L165) seems to resolve the original issue. Would this block the isolate until the flush completes? If yes, can a separate thread wait out the flush and close the handle when `FlushFileBuffers` return?

**Comment by brianquinlan** [comment]
@mtolmacs I need to do some more research - `FlushFileBuffers` on `stdin` will block forever if the child process doesn't read `stdin`.

**Comment by mraleph** [comment]
I would think that Linux pipes likely have similar behaviour though we don't hit it because default pipe buffer is larger. If you write to a pipe and the other side does not read you will eventually block on writing (blocking IO) or you would be told to try again (non-blocking IO). 

FWIW I think we should probably change how we implement `flush` to try and guarantee the invariant that all data is written out into kernel (does not mean that kernel has delivered the data yet, but at least this data is not going to disappear). Even though it would mean that something like `flush` would potentially block indefinitely.

**Comment by brianquinlan** [comment]
I'm not sure that we need to get the operating system to flush.

It turns out that, on Windows, if the `WriteFile` is larger than the buffer size then `CloseHandle` is called before the the `WriteFile` completes. If I call `CloseHandle` the instant that `WriteFile` succeeds then the subprocess can read the data.

I'm trying to figure out who gets notified when the write completes.

**Comment by brianquinlan** [comment]
Contrary to my understanding of the code, `flush` does work on macOS (though it is very slow for the 128MB buffer that I tried). I need to look at the code again.

**Comment by brianquinlan** [comment]
OK, I think that I have a better understanding of what is going on now.

`_SocketStreamConsumer.write()` writes as much of the buffer as it can:

```dart
  void write() {
    final sub = subscription;
    if (sub == null) return;
    // Write as much as possible.
    print('write() - offset: $offset ${buffer!.length} ');
    offset =
        offset! + socket._write(buffer!, offset!, buffer!.length - offset!);
    print('  offset: $offset');
    if (offset! < buffer!.length) {
      if (!paused) {
        paused = true;
        sub.pause();
      }
      socket._enableWriteEvent();
    } else {
      buffer = null;
      if (paused) {
        paused = false;
        sub.resume();
      }
    }
  }
```

On POSIX, the pipe representing stdin is set to not block. So when the socket is written to like:

```dart
intptr_t SocketBase::Write(intptr_t fd,
                           const void* buffer,
                           intptr_t num_bytes,
                           SocketOpKind sync) {
  printf("SocketBase::Write(%ld, %p, %ld, %d)\n", fd, buffer, num_bytes, sync);
  ASSERT(fd >= 0);
  ssize_t written_bytes = TEMP_FAILURE_RETRY(write(fd, buffer, num_bytes));
  ASSERT(EAGAIN == EWOULDBLOCK);
  if ((sync == kAsync) && (written_bytes == -1) && (errno == EWOULDBLOCK)) {
    // If the would block we need to retry and therefore return 0 as
    // the number of bytes written.
    written_bytes = 0;
  }
  return written_bytes;
}
```

So we write as much of the buffer out as possible without blocking (possibly 0 bytes) and then try to write more when the socket (really a pipe) says that it is ready to write.

On Windows, we using `WriteFile` in overlapped mode, which can block internally if the buffer is too small. So `_SocketStreamConsumer.write()` isn't actually tracking what has been written.

**Comment by brianquinlan** [comment]
I tried modifying `_SocketStreamConsumer.write()` to not attribute the write until receiving a [`write` `SocketEvent`](https://api.dart.dev/stable/2.19.2/dart-io/RawSocketEvent/write-constant.html):

```dart
  void write() {
    offset = offset! + _lastWriteSize;
    final sub = subscription;
    if (sub == null) return;

    if (offset! < buffer!.length) {
      // Write as much as possible.
      _lastWriteSize = socket._write(buffer!, offset!, buffer!.length - offset!);

      if (!paused) {
        paused = true;
        sub.pause();
      }
      socket._enableWriteEvent();
    } else {
      buffer = null;
      _lastWriteSize = 0;
      if (paused) {
        paused = false;
        sub.resume();
      }
    }
  }
```

But a write event is sent **before** the write actually completes. I'm not sure if that is correct or not.

Changing the PIPE buffer size to [`kBufferSize`](https://github.com/dart-lang/sdk/blob/3571b97b24a2d8b27038284d5f0be58bd0733f61/runtime/bin/eventhandler_win.cc#L33) (the maximum amount that we will write in a single system call) fixes the issue on my machine but it seems brittle.

**Comment by brianquinlan** [comment]
Increasing the buffer size doesn't always fix the issue:
https://dart-review.googlesource.com/c/sdk/+/284487

**Comment by mraleph** [comment]
My interpretation so far is that if we want the data to be reliably delivered we need to accept that flushing operation might need to block.

**Comment by loic-sharma** [comment]
Do you have any recommendations on how one can work around this bug? This blocks supporting Flutter's pre-push checks on Windows: https://github.com/flutter/engine/pull/36123

**Comment by brianquinlan** [comment]
Sorry, I'm back to working on this. Sorry for the delay. It will take me a while to catch back up on this.

On possible hacky workaround to this would be to do a small write (e.g. 1K) and then a delay to allow the receiving process time to read the data, e.g., by delaying 0.1s. Then write the next bit of data and then so on.

**Comment by brianquinlan** [comment]
I changed:

```c
intptr_t Handle::Write(const void* buffer, intptr_t num_bytes) {
  MonitorLocker ml(&monitor_);
  if (HasPendingWrite()) {
    return 0;
  }
  if (num_bytes > kBufferSize) {
    num_bytes = kBufferSize;
  }
  ASSERT(SupportsOverlappedIO());
  if (completion_port_ == INVALID_HANDLE_VALUE) {
    return 0;
  }
  int truncated_bytes = Utils::Minimum<intptr_t>(num_bytes, INT_MAX);
  pending_write_ = OverlappedBuffer::AllocateWriteBuffer(truncated_bytes);
  pending_write_->Write(buffer, truncated_bytes);
  if (!IssueWrite()) {
    return -1;
  }

+  FlushFileBuffers(handle());

  return truncated_bytes;
}
```

and it solved the issue.  Making that call in a synchronous context e.g. during `flush` seems like the simplest approach.

**Comment by mraleph** [comment]
I think this will do the trick: https://dart-review.googlesource.com/c/sdk/+/307043

**Comment by mtolmacs** [comment]
I tried the patch with various content lengths and it indeed seems to work. How can this be tracked outside of googlesource.com? I'd love to know which version of Dart this fix will be merged into, so I can track it on the Flutter side as well. Thx!
